# --- Day 2: Red-Nosed Reports ---

Fortunately, the first location The Historians want to search isn't a long walk from the Chief Historian's office.

While the Red-Nosed Reindeer nuclear fusion/fission plant appears to contain no sign of the Chief Historian, the engineers there run up to you as soon as they see you. Apparently, they still talk about the time Rudolph was saved through molecular synthesis from a single electron.

They're quick to add that - since you're already here - they'd really appreciate your help analyzing some unusual data from the Red-Nosed reactor. You turn to check if The Historians are waiting for you, but they seem to have already divided into groups that are currently searching every corner of the facility. You offer to help with the unusual data.

The unusual data (your puzzle input) consists of many reports, one report per line. Each report is a list of numbers called levels that are separated by spaces. For example:

```
7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9
```

This example data contains six reports each containing five levels.

The engineers are trying to figure out which reports are safe. The Red-Nosed reactor safety systems can only tolerate levels that are either gradually increasing or gradually decreasing. So, a report only counts as safe if both of the following are true:

- The levels are either all increasing or all decreasing.
- Any two adjacent levels differ by at least one and at most three.

In the example above, the reports can be found safe or unsafe by checking those rules:

```
7 6 4 2 1: Safe because the levels are all decreasing by 1 or 2.
1 2 7 8 9: Unsafe because 2 7 is an increase of 5.
9 7 6 2 1: Unsafe because 6 2 is a decrease of 4.
1 3 2 4 5: Unsafe because 1 3 is increasing but 3 2 is decreasing.
8 6 4 4 1: Unsafe because 4 4 is neither an increase or a decrease.
1 3 6 7 9: Safe because the levels are all increasing by 1, 2, or 3.
```

So, in this example, 2 reports are safe.

Analyze the unusual data from the engineers. How many reports are safe?

In [11]:
def read_input(file_name):
    input = []
    with open(file_name, 'r') as file:
        for line in file:
            input.append([int(x) for x in line.split()])
    return input

In [12]:
def safe(levels) -> bool:
    # print(levels)
    if levels[0] == levels[1]:
        # print('Unsafe equal value')
        return False

    ascending = levels[0] < levels[1]
    
    for i in range(len(levels) -1):
        diff = levels[i] - levels[i+1]
        # print(f'{ascending=} {diff} = {levels[i]} - {levels[i+1]}')
        if ascending and diff not in [-1,-2,-3] or (not ascending and diff not in [1,2,3]):
            # print('Unsafe rapid change')
            return False
        
    # print('Safe!')    
    return True
            

In [13]:
def count_iterable(i):
    return sum(1 for e in i)

In [14]:
levels_list = read_input('input.txt')
safe_levels = filter(safe, levels_list)
print(count_iterable(safe_levels))

559


Your puzzle answer was 559.

The first half of this puzzle is complete! It provides one gold star: *

# --- Part Two ---
The engineers are surprised by the low number of safe reports until they realize they forgot to tell you about the Problem Dampener.

The Problem Dampener is a reactor-mounted module that lets the reactor safety systems tolerate a single bad level in what would otherwise be a safe report. It's like the bad level never happened!

Now, the same rules apply as before, except if removing a single level from an unsafe report would make it safe, the report instead counts as safe.

More of the above example's reports are now safe:

```
7 6 4 2 1: Safe without removing any level.
1 2 7 8 9: Unsafe regardless of which level is removed.
9 7 6 2 1: Unsafe regardless of which level is removed.
1 3 2 4 5: Safe by removing the second level, 3.
8 6 4 4 1: Safe by removing the third level, 4.
1 3 6 7 9: Safe without removing any level.
```

Thanks to the Problem Dampener, 4 reports are actually safe!

Update your analysis by handling situations where the Problem Dampener can remove a single level from unsafe reports. How many reports are now safe?

In [15]:
def find_asc_or_desc(levels):
    for i in range(len(levels) -1):
        if levels[i] - levels[i+1] > 0:
            return False
        elif levels[i] - levels[i+1] < 0:
            return True
    return None

In [16]:
def problem_levels(levels):
    
    ascending = find_asc_or_desc(levels)
    # print(f'{ascending=}')
    
    problems = []
    
    for i in range(len(levels) -1):
        diff = levels[i] - levels[i+1]
        # print(f'{ascending=} {diff} = {levels[i]} - {levels[i+1]}')
        if ascending and diff not in [-1,-2,-3] or (not ascending and diff not in [1,2,3]):
            problems.append((i))
        
    # print(f'{problems=}')    
    return problems

In [17]:
def remove_problem(levels, p):
    new_list = levels.copy()
    new_list.pop(p)
    return safe(new_list)

In [18]:
def check_dampener_options(levels, problems): 
    if len(problems) == 0: 
        return True

    for p in problems:
        if remove_problem(levels, p):
            return True
        if remove_problem(levels, p+1):
            return True
        
    return False

In [19]:
def safe_with_dampener(levels_list, problems):
    safe_count = 0
    for i in range(len(levels_list)):
        # print(f'{levels_list[i]=}    {problems[i]=}')
        if check_dampener_options(levels_list[i], problems[i]):
            safe_count += 1
    return safe_count

In [20]:
problems = [problem_levels(l) for l in levels_list]
print(safe_with_dampener(levels_list, problems))

601
